tools/licenses/lib/licenses.dart

```diff
@@ -116,21 +153,132 @@ LicenseType convertBodyToType(String body) {
 }
 
 abstract class LicenseSource {
+  String get name;
+  String get libraryName;
+  String get officialSourceLocation;
   List<License>? nearestLicensesFor(String name);
   License? nearestLicenseOfType(LicenseType type);
   License? nearestLicenseWithName(String name, { String? authors });
 }
 
-abstract class License implements Comparable<License> {
+class Assignment {
+  const Assignment(this.license, this.target, this.source);
+  final License license;
+  final String target;
+  final LicenseSource source;
+}
+
+class GroupedLicense {
+  GroupedLicense(this.type, this.body);
+  final LicenseType type;
+  final String body;
+  final Set<String> targets = <String>{};
+  final Set<String> libraries = <String>{};
+  final Set<String> origins = <String>{};
+
+  String toStringDebug() {
+    final StringBuffer result = StringBuffer();
+    result.writeln('=' * 100);
+    (libraries.map((String s) => 'LIBRARY: $s').toList()..sort()).forEach(result.writeln);
+    (origins.map((String s) => 'ORIGIN: $s').toList()..sort()).forEach(result.writeln);
+    result.writeln('TYPE: $type');
+    (targets.map((String s) => 'FILE: $s').toList()..sort()).forEach(result.writeln);
+    result.writeln('-' * 100);
+    if (body.isEmpty) {
+      result.writeln('<THIS BLOCK INTENTIONALLY LEFT BLANK>');
+    } else {
+      result.writeln(body);
+    }
+    result.writeln('=' * 100);
+    return result.toString();
+  }
+
+  String toStringFormal() {
+    final StringBuffer result = StringBuffer();
+    (libraries.toList()..sort()).forEach(result.writeln);
+    result.writeln();
+    assert(body.isNotEmpty);
+    result.write(body);
+    return result.toString();
+  }
+}
+
+/*
+// TODO(ianh): Remove this after the PR lands, to reorder the licenses to the new order.
```

Comment:
Is this supposed to be uncommented in the PR?

Reply:
oh right i can just remove this now. good catch.